### PR TITLE
Encrypt votes

### PR DESCRIPTION
**Pull request by KonradSwierczynski**
Send votes from frontend to backend encrypted using election's public key.
The encryption used: libsodium Sealedbox encryption (https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes)

TLDR it's using authenticated encryption but as the authentication is not needed (user so far has any key) the sender keys are just temporary values, and the nonce is generated based on the sender public key (which is random)

The frontend lib tweetnacl doesn't provide Sealedbox, but it is the matter of nonce generation, which I did myself basing on https://github.com/whs/tweetnacl-sealed-box

### CODE REVIEW

File: backend/apollo/elections/views.py [hunk 1]
```diff
@@ -99,6 +101,34 @@ def get_permissions(self) -> List[BasePermission]:
...
+class VoteViewSet(viewsets.GenericViewSet):
     queryset = Vote.objects.all()
     serializer_class = VoteSerializer
+
+    def create(self, request):
+        try:
+            answer_id_ciphertext = request.data["answer"]
+            election_id = request.data["election"]
+        except KeyError as e:
+            return Response(f"Key {e} not found", status=status.HTTP_400_BAD_REQUEST)
```
Comment (adambudziak) [inline backend/apollo/elections/views.py:113]: I believe you could override `create()` inside `VoteSerializer`. Then you could validate the request using `validate()` as usual

File: backend/apollo/elections/views.py [hunk 2]
```diff
@@ -99,6 +101,34 @@ def get_permissions(self) -> List[BasePermission]:
...
+                election.secret_key.encode("ascii"), encoder
+            )
+            sealed_box = nacl.public.SealedBox(election_sk)
+            answer_id = sealed_box.decrypt(answer_id_ciphertext, encoder).decode()
+        except nacl.exceptions.CryptoError:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        answer = get_object_or_404(Answer, pk=answer_id)
+        if answer.question.election.id != election_id:
+            return Response("Wrong election id", status=status.HTTP_400_BAD_REQUEST)
```
Comment (adambudziak) [inline backend/apollo/elections/views.py:129]: So everything above this line could be in `validate()` in the serializer, and in the last line of this method you'd just set
```python
def validate(attrs):
    attrs = super().validate(attrs)
    ...  # all the crypto stuff
    attrs["answer"] = get_object_or_404(Answer, pk=answer_id)
    return attrs
```

File: backend/apollo/elections/serializers.py [hunk 3]
```diff
@@ -55,17 +59,53 @@ def authorize(self, attrs):
...
+        answer = get_object_or_404(Answer, pk=answer_id)
+
+        if answer.question.election.id != election_id:
+            if raise_exception:
+                raise ValidationError()
+            return False
+
+        self.initial_data["answer"] = answer_id
+
+        return super().is_valid(raise_exception)
```
Comment (adambudziak) [inline backend/apollo/elections/serializers.py:102]: How about this?

```python
    def _decrypt_election(self):
        election = get_object_or_404(Election, pk=self.initial_data.pop("election"))
        answer_id = decrypt(election.secret_key, self.initial_data["answer"])
        answer = get_object_or_404(Answer, pk=answer_id)

        if answer.question.election.id != election_id:
            raise ValidationError()

        return answer_id

    def is_valid(self, raise_exception):
        try:
            answer_id = self._decrypt_election()
        except (KeyError, CryptoError, ValidationError):
            if raise_exception:
                raise ValidationError()

        self.initial_data["answer"] = answer_id
        return super().is_valid(raise_exception)

```
With `pop` in the first line you don't need the `create` method.

In the future we can think about doing it even prettier.

File: frontend/src/views/VotePage.vue [hunk 4]
```diff
@@ -34,14 +35,17 @@ export default Vue.extend({
     voteInElection: function (votes: Array<Vote>) {
       const results = votes.map((vote) => {
         console.log(vote);
-        if (vote.selected === null) {
-          console.error("Vote is null");
-          return Promise.reject("Vote is null");
+        if (vote.selected === null || this.election === null) {
+          console.error("Vote or election is null");
```
Comment (adambudziak) [inline frontend/src/views/VotePage.vue:39]: `console.error` will not deploy to prod

File: backend/apollo/elections/crypto.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    try:
+        election_sk = nacl.public.PrivateKey(
+            election_secret_key.encode("ascii"), encoder
+        )
+        sealed_box = nacl.public.SealedBox(election_sk)
+        message = sealed_box.decrypt(ciphertext, encoder).decode()
+    except TypeError as e:
+        raise
+    except nacl.exceptions.CryptoError as e:
+        raise CryptoError from e
```
Comment (adambudziak) [inline backend/apollo/elections/crypto.py:19]: That's some wild error handling.

File: backend/apollo/elections/serializers.py (hunk 3 above)
Comment (KonradSwierczynski) [reply backend/apollo/elections/serializers.py:102]: That is much better, thanks

**Pull request merged**